src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -853,132 +874,185 @@ public void ProcessQueue(SocketAsyncContext context)
+                            Debug.Assert(_tail == null);
+                            Trace(context, $"Exit (stopped)");
+                            return OperationResult.Cancelled;
+                        }
+                        else
+                        {
+                            Debug.Assert(_state == QueueState.Processing, $"_state={_state} while processing queue!");
+
+                            if (observedSequenceNumber != _sequenceNumber)
+                            {
+                                // We received another epoll notification since we previously checked it.
+                                // So, we need to retry the operation.
+                                Debug.Assert(observedSequenceNumber - _sequenceNumber < 10000, "Very large sequence number increase???");
+                                observedSequenceNumber = _sequenceNumber;
+                            }
+                            else
+                            {
+                                _state = QueueState.Waiting;
+                                Trace(context, $"Exit (received EAGAIN)");
+                                return OperationResult.Pending;
+                            }
+                        }
+                    }
+                }
+
+                // Remove the op from the queue and see if there's more to process.
+
+                AsyncOperation nextOp = null;
+                using (Lock())
+                {
+                    if (_state == QueueState.Stopped)
+                    {
+                        Debug.Assert(_tail == null);
+                        Trace(context, $"Exit (stopped)");
+                    }
+                    else
                     {
-                        // Try to perform the IO
-                        wasCompleted = op.TryComplete(context);
-                        if (wasCompleted)
+                        Debug.Assert(_state == QueueState.Processing, $"_state={_state} while processing queue!");
+                        Debug.Assert(_tail.Next == op, "Queue modified while processing queue");
+
+                        if (op == _tail)
                         {
-                            needCallback = op.SetComplete();
+                            // No more operations to process
+                            _tail = null;
+                            _state = QueueState.Ready;
+                            _sequenceNumber++;
+                            Trace(context, $"Exit (finished queue)");
                         }
                         else
                         {
-                            op.SetWaiting();
+                            // Pop current operation and advance to next
+                            nextOp = _tail.Next = op.Next;
                         }
                     }
+                }
+
+                if (nextOp != null)
+                {
+                    nextOp.Dispatch(s_processingCallback);
+                }
```

Comment:
Nit: could just be `nextOp?.Dispatch(s_processingCallback);`, but this is fine, too